### PR TITLE
Approximate indirect specular occlusion

### PR DESCRIPTION
**Pull request by aevyrie**
# Objective

- The current PBR renderer over-brightens indirect specular reflections, which tends to cause objects to appear to glow, because specular occlusion is not accounted for.

## Solution

- Attenuate indirect specular term with an approximation for specular occlusion, using [[Lagarde et al., 2014] (pg. 76)](https://seblagarde.files.wordpress.com/2015/07/course_notes_moving_frostbite_to_pbr_v32.pdf).

| Before | After | Animation |
| --- | --- | --- |
| before bike | after bike | ezgif-1-fbcbaf272b |
| classroom before | classroom after | ezgif-1-0f390edd06 |

---

## Changelog

- Ambient occlusion now applies to indirect specular reflections to approximate how objects occlude specular light.

## Migration Guide

- Renamed `PbrInput::occlusion` to `diffuse_occlusion`, and added `specular_occlusion`.

### CODE REVIEW

**Review by IceSentry — COMMENTED** [review]
I really like how it looks especially considering how small the change is but since it breaks the Physically Based part we should probably make this configurable.

The suggestion in discord to do `indirect_light += environment_light.diffuse * occlusion + environment_light.specular * specular_occlusion` would make that pretty easy.

The comment should maybe be shortened a bit, maybe don't mention the wheel well example? It seems a bit too specific and over explained to me.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:326]:
```suggestion
    // This is technically not physically correct. Ambient occlusion should only attenuate the
```

**Comment by aevyrie** [comment]
> it breaks the Physically Based part

I think this is debatable. Not accounting for specular occlusion is also not physical, but it requires raytracing to simulate accurately. This gets us closer to the ground truth by having *something* for the specular occlusion term. The next most accurate would be using SSR for the occlusion check, followed by full path tracing.

**Review by nicopap — APPROVED** [review]
I do agree that it looks better when occluding the specular as well. The Fresnel defeats pretty much the purpose of the occlusion.

I followed closely the discussion on discord I think this is worth merging.

I've yet to see a scene where it looks worse, so I think it should be merged, since it's not even a tradeoff, it's just straight up a positive change.

Of note is that `occlusion` is not strictly defined by SSAO, but also by the occlusion texture. But looking [at the glTF2 spec](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#additional-textures) I don't see any details about whether this should affect environment light specular reflection or not, so it's fine as-is.

Regarding the suggestion of having it parametrable. I think it should be part of a different PR. It's a different level of controversy, as it includes more tradeoffs.

It seems @DGriffin91 has a more accurate implementation of specular occlusion, but until its ready, I think it's fine to merge this.

**Comment by DGriffin91** [comment]
Specular ambient occlusion [is included](https://github.com/GameTechDev/XeGTAO/blob/master/Source/Rendering/Shaders/Filament/light_indirect.va.fs#L502) in both reference implementation that bevy uses for GTAO ([XeGTAO](https://github.com/GameTechDev/XeGTAO)), and is also specified in original paper: [[Jimenez et al., 2016] (pg. 3)](https://www.activision.com/cdn/research/Practical_Real_Time_Strategies_for_Accurate_Indirect_Occlusion_NEW%20VERSION_COLOR.pdf)

Specular occlusion from [[Lagarde et al., 2014] (pg. 76)](https://seblagarde.files.wordpress.com/2015/07/course_notes_moving_frostbite_to_pbr_v32.pdf) is also specified in [Filament](https://google.github.io/filament/), the primary reference PBR implementation that bevy uses: https://google.github.io/filament/Filament.md.html#lighting/occlusion/specularocclusion

(XeGTAO implements both the simpler Lagarde14 method as well as the improved Jimenez16 one)

Using the diffuse ambient occlusion term directly for specular occlusion will result in over occlusion in areas that would receive less ambient diffuse light but still have a path for specular light, while also significantly under occluding in most cases.

We should continue implementing GTAO to include GTSO. And then also provide a SSR oriented method for more accurate specular occlusion that also better captures more distant occlusion.

**Comment by aevyrie** [comment]
The filament docs and related paper you link, are equivalent to what I've done here, for rough materials. For rough materials, AO and specular occlusion are 1:1.

File: crates/bevy_pbr/src/deferred/deferred_lighting.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/deferred/deferred_lighting.wgsl:75]:
```suggestion
        // Use SSAO to estimate the specular occlusion. [Lagarde et al., 2014]
```
Nit

File: crates/bevy_pbr/src/deferred/deferred_lighting.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/deferred/deferred_lighting.wgsl:76]: Does it make sense to use only the SSAO here, and not min(ssao, baked_ao)? No idea which is better.

For context, the baked AO should be treated kinda like a normal map. It's _not_ for occlusion between multiple meshes (that's what lightmaps are for). It's instead occlusion baked as part of the material, such as micro-cracks and indents in the material.

We should update the docs to make this a bit more explicit imo.

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/render/pbr_fragment.wgsl:177]:
```suggestion
        // Use SSAO to estimate the specular occlusion. [Lagarde et al., 2014]
```

**Review by JMS55 — APPROVED** [review]
Needs a migration guide and changelog for the diffuse_occlusion naming change.

Approving this PR, assuming that ignoring baked occlusion makes sense.

File: crates/bevy_pbr/src/deferred/deferred_lighting.wgsl
Comment (aevyrie) [reply crates/bevy_pbr/src/deferred/deferred_lighting.wgsl:76]: The baked AO is a vec3, I'm not sure how to convert that to a visibility value, maybe luminance?

Comment (JMS55) [reply crates/bevy_pbr/src/deferred/deferred_lighting.wgsl]: The baked occlusion is a single f32. Check where the texture is read in PBR fragment, we only read the R channel. It's simply cast to a vec3 for compatibility with SSAO multibounce.

Comment (aevyrie) [reply crates/bevy_pbr/src/deferred/deferred_lighting.wgsl]: Thanks, I missed that.

File: crates/bevy_pbr/src/deferred/deferred_lighting.wgsl
Comment (IceSentry) [inline crates/bevy_pbr/src/deferred/deferred_lighting.wgsl:75]: Maybe add a link to the paper or the title of the paper?

**Review by IceSentry — APPROVED** [review]

**Comment by aevyrie** [comment]
Note, I implemented a `min(baked_ao, ssao)` for specular, but this had no effect because SSAO was always darker. We decided to just ignore the baked AO for specular occlusion.

File: crates/bevy_pbr/src/deferred/deferred_lighting.wgsl
Comment (aevyrie) [reply crates/bevy_pbr/src/deferred/deferred_lighting.wgsl]: We attempted this and decided to revert.

**Review by JMS55 — APPROVED** [review]

**Review by nicopap — APPROVED** [review]

**Review by rodolphito — APPROVED** [review]

**Pull request merged**
